includes/abilities-api/class-wp-abilities-registry.php

```diff
@@ -159,17 +147,23 @@ public function register( $name, array $properties = array() ): ?WP_Ability {
 			return null;
 		}
 
-		$ability = new WP_Ability(
+		/**
+		 * Filters the class used to instantiate the ability.
+		 *
+		 * This is helpful for creating complex ability types that are not well expressed by array values and callbacks.
+		 *
+		 * @param string $ability_class           The class name to instantiate the ability. Must extend \WP_Ability.
+		 * @param string $name                    The name of the ability being registered.
+		 * @param array<string,mixed> $properties The properties of the ability being registered. See
+		 *                                        `wp_register_ability()` for the expected structure.
+		 *
+		 * @phpstan-param class-string<\WP_Ability> $ability_class
+		 */
+		$ability_class = apply_filters( 'wp_ability_class', WP_Ability::class, $name, $properties );
+
+		$ability = new $ability_class(
 			$name,
-			array(
-				'label'               => $properties['label'],
-				'description'         => $properties['description'],
-				'input_schema'        => $properties['input_schema'] ?? array(),
-				'output_schema'       => $properties['output_schema'] ?? array(),
-				'execute_callback'    => $properties['execute_callback'],
-				'permission_callback' => $properties['permission_callback'] ?? null,
-				'meta'                => $properties['meta'] ?? array(),
-			)
+			$properties
```

Comment:
It's fine to remove this mapping, but we will have to ensure that in `WP_Ability::__construct` we only assign properties that exist using `property_exists()` or something similar.

Reply:
Currently, if the property doesn't exist then the constructor will throw an error (unless the user has extended the class to add specific or `#AllowDynamicProperties`)
- https://github.com/WordPress/abilities-api/blob/2da7f4630a30c055b089a6e33a66fb5222a297d2/includes/abilities-api/class-wp-ability.php#L116

Will switch to a `_doing_it_wrong()` and `continue` when I get to my desk.

Reply:
Done in ddd5786688124e07474bc38de36aeea75c8767c3